File: src/loom/api/catalog/lifecycle.py
from pydantic import BaseModel

from loom.domain.enums import LifecycleState

LEGAL_TRANSITIONS: dict[LifecycleState, frozenset[LifecycleState]] = {
    LifecycleState.DRAFT: frozenset({LifecycleState.IN_REVIEW}),
    LifecycleState.IN_REVIEW: frozenset(
        {LifecycleState.APPROVED, LifecycleState.DRAFT}
    ),
    LifecycleState.APPROVED: frozenset({LifecycleState.PUBLISHED}),
    LifecycleState.PUBLISHED: frozenset({LifecycleState.DEPRECATED}),
    LifecycleState.DEPRECATED: frozenset({LifecycleState.RETIRED}),
    LifecycleState.RETIRED: frozenset(),
}


def is_legal_transition(current: LifecycleState, target: LifecycleState) -> bool:
    """Check whether target is a legal next state from current."""
    return target in LEGAL_TRANSITIONS.get(current, frozenset())


class TransitionRequest(BaseModel):
    """Request body for a lifecycle-transition endpoint, shared by all routers."""

    to_state: LifecycleState

File: clients/vscode/src/lifecycle.ts
// Mirrors `src/loom/domain/lifecycle.py`'s `LEGAL_TRANSITIONS` -- the same
// state machine every versioned entity shares:
//   Draft -> In Review -> Approved -> Published -> Deprecated -> Retired
// The server is still the authority (RBAC scope `catalog:*:transition` plus
// eval gates it can reject on); this table only keeps the editor from
// offering a target state that's structurally impossible.

export const LIFECYCLE_STATES = [
  'draft',
  'in_review',
  'approved',
  'published',
  'deprecated',
  'retired',
] as const;

export type LifecycleState = (typeof LIFECYCLE_STATES)[number];

const LEGAL_TRANSITIONS: Record<LifecycleState, LifecycleState[]> = {
  draft: ['in_review'],
  in_review: ['approved', 'draft'],
  approved: ['published'],
  published: ['deprecated'],
  deprecated: ['retired'],
  retired: [],
};

export function legalTransitions(from: string): LifecycleState[] {
  return LEGAL_TRANSITIONS[from as LifecycleState] ?? [];
}

/** Human-facing label for a state, e.g. `in_review` -> "In Review". */
export function lifecycleLabel(state: string): string {
  return state
    .split('_')
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}
